cdap-watchdog/src/main/java/co/cask/cdap/logging/meta/AbstractCheckpointManager.java

```diff
@@ -0,0 +1,142 @@
+    if (lastCheckpoint.equals(checkpoints)) {
+      return;
+    }
+
+    lastCheckpoint = TransactionRunners.run(transactionRunner, context -> {
+      Map<Integer, Checkpoint<T>> result = new HashMap<>();
+      StructuredTable table = context.getTable(StoreDefinition.LogCheckpointStore.LOG_CHECKPOINT_TABLE);
+      for (Map.Entry<Integer, ? extends Checkpoint<T>> entry : checkpoints.entrySet()) {
+        Checkpoint<T> checkpoint = entry.getValue();
+        Collection<Field<?>> fields = getKeyFields(rowKeyPrefix, entry.getKey());
+        fields.add(Fields.bytesField(StoreDefinition.LogCheckpointStore.CHECKPOINT_FIELD,
+                                     serializeCheckpoint(checkpoint)));
+        table.upsert(fields);
+        result.put(entry.getKey(), new Checkpoint<>(checkpoint.getOffset(), checkpoint.getMaxEventTime()));
+      }
+      return result;
+
+    }, IOException.class);
+
+    LOG.trace("Saved checkpoints for partitions {}", checkpoints);
+  }
+
+  @Override
+  public Map<Integer, Checkpoint<T>> getCheckpoint(Set<Integer> partitions) throws IOException {
+    return TransactionRunners.run(transactionRunner, context -> {
+      Map<Integer, Checkpoint<T>> checkpoints = new HashMap<>();
+      StructuredTable table = context.getTable(StoreDefinition.LogCheckpointStore.LOG_CHECKPOINT_TABLE);
+      for (int partition : partitions) {
+        Optional<StructuredRow> optionalRow = table.read(getKeyFields(rowKeyPrefix, partition));
+        StructuredRow row = optionalRow.orElse(null);
+        checkpoints.put(partition, fromRow(row));
+      }
+      return checkpoints;
+    }, IOException.class);
+  }
+
+  @Override
+  public Checkpoint<T> getCheckpoint(int partition) throws IOException {
+    Checkpoint<T> checkpoint = TransactionRunners.run(transactionRunner, context -> {
+      StructuredTable table = context.getTable(StoreDefinition.LogCheckpointStore.LOG_CHECKPOINT_TABLE);
+      Optional<StructuredRow> optionalRow = table.read(getKeyFields(rowKeyPrefix, partition));
+      StructuredRow row = optionalRow.orElse(null);
+      return fromRow(row);
+    }, IOException.class);
+
+    LOG.trace("Read checkpoint {} for partition {}", checkpoint, partition);
+    return checkpoint;
+  }
+
+  private Checkpoint<T> fromRow(@Nullable StructuredRow row) throws IOException {
+    byte[] offset = null;
+    if (row != null) {
+      offset = row.getBytes(StoreDefinition.LogCheckpointStore.CHECKPOINT_FIELD);
+    }
+    return deserializeCheckpoint(offset);
+  }
+
+  /**
+   * Serialize the checkpoint.
+   * @param checkpoint checkpoint to be serialized
+   * @return serialized checkpoint byte array
+   * @throws IOException if error while serializing checkpoint
+   */
+  protected abstract byte[] serializeCheckpoint(Checkpoint<T> checkpoint) throws IOException;
+
+  /**
+   * Deserialize the checkpoint bytes. If checkpoint bytes are null, then return {@code -1} for all the numeric
```

Comment:
This is because `CheckpointManager` expects -1 as value if checkpoint is missing: https://github.com/cdapio/cdap/blob/develop/cdap-watchdog/src/main/java/co/cask/cdap/logging/meta/CheckpointManager.java#L35

Reply:
And then why the `CheckpointManager` requires `-1`?

Reply:
Remove that from contract.